Adds a Gyroscope Sensor.

liangfok opened the pull request:
This is a System 2.0 port of `RigidBodySystem::RigidBodyGyroscope`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4795)
<!-- Reviewable:end -->

siyuanfeng-tri commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagisYbPZR3AqFio7wg:-KagisYbPZR3AqFio7wh:b-19piyv) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.cc#L21)):*
> ```cpp
>                      const RigidBodyTree<double>& tree)
>     : name_(name),
>       frame_(frame),
> ```

i believe in the current implementation, this @p frame has to be owned by this rigidbody tree. there is nothing wrong with this assumption, but it's probably better to state it in the comments. once #4590 is in, we can switch to the CalcFrameVelocity to lift this assumption. 

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KaghlvE31w43iJESYaj:-KaghlvE31w43iJESYak:bheluee) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.cc#L66)):*
> ```cpp
>   Eigen::Vector3d angular_rates = relative_twist.head<3>();
> 
>   // Saves the acceleration readings into the output port.
> ```

velocity? you are outputting angular velocity here. 

---

*[drake/systems/sensors/gyroscope.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagiMbMYVyBjDkpZ6sc:-KagiMbMYVyBjDkpZ6sd:be88lnh) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.h#L21)):*
> ```objc
> namespace sensors {
> 
> /// A simulated ideal gyroscope that measures angular acceleration.
> ```

velocity?

---

*[drake/systems/sensors/gyroscope.h, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagidLmiiZqtwPj-LYW:-KagidLmiiZqtwPj-LYX:b-am0p55) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.h#L27)):*
> ```objc
> /// There are two input ports that are accessible via the following accessors:
> ///
> ///  - get_state_input_port(): Contains the state (i.e., postion and velocity)
> ```

these are outdated? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-cpplint-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-xenial-gcc-experimental-open-source please

liangfok commented:
+@siyuanfeng-tri for feature review please. Thanks!

---

Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagisYbPZR3AqFio7wg:-KcaS-d5GZ0AANizGeeN:b-yohi4k) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.cc#L21)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i believe in the current implementation, this @p frame has to be owned by this rigidbody tree. there is nothing wrong with this assumption, but it's probably better to state it in the comments. once #4590 is in, we can switch to the CalcFrameVelocity to lift this assumption. 
</blockquote></details>

Done. I updated the documentation to indicate the frame need not be in the tree, but must at least reference a body in the tree.

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KaghlvE31w43iJESYaj:-KbzHZTmHLY1k-2-Fox1:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.cc#L66)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

velocity? you are outputting angular velocity here. 
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagiMbMYVyBjDkpZ6sc:-KbzINZe7oxhd_4rEujW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.h#L21)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

velocity?
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KagidLmiiZqtwPj-LYW:-KbzIWy6-TEHdHNqB5y2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/54d5f89127b39fb8c629668c8db1648079fcd536/drake/systems/sensors/gyroscope.h#L27)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

these are outdated? 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-Kcc1D7wWZ-FIfWl35iz:bex8s86)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccIWGAePOy3WtuB_Kg:-KccIWGAePOy3WtuB_Kh:b5geyuc) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.cc#L61)):*
> ```cpp
>       tree_.CalcFramePoseInWorldFrame(cache, frame_);
> 
>   // Extracts inverse rotation matrix from transform, and translational
> ```

no translational acc? 

---

*[drake/systems/sensors/gyroscope.h, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccImD03mgsGS93zmoz:-KccImD1fNhLwzTi3rke:b-neq2et) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L29)):*
> ```objc
> ///  - `G` be the gyroscope's frame with origin point `Go`.
> ///  - `W` be the world frame with origin point `Wo`.
> ///  - `ω_WGo_W` be `Go`'s rotational velocity in `W` expressed in `W`.
> ```

G's rotational velocity, Go is a point, and it is kinda weird to say it angular velocity. Also everything rigidly attached to G has the same angular velocity. 

---

*[drake/systems/sensors/gyroscope.h, line 31 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJ1bvaK96m3CQxvcH:-KccJ1bwCZe76rJvoyht:b3krm) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L31)):*
> ```objc
> ///  - `ω_WGo_W` be `Go`'s rotational velocity in `W` expressed in `W`.
> ///  - `R_GW` be the rotation matrix from `W` to `G`.
> ///  - `ω_WGo_G` be `Go`'s rotational velocity in `W` expressed in `G`. This is
> ```

G

---

*[drake/systems/sensors/gyroscope.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJJi9jgsOGZw-EheA:-KccJJi9jgsOGZw-EheB:b-38n0bh) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L79)):*
> ```objc
>       const RigidBodyTree<double>& tree);
> 
>   // Non-copyable.
> ```

BTW, should we switch to drake_copyable? there is a macro for something like this. 

---

*[drake/systems/sensors/gyroscope_output.cc, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJkaZSH2yeilFy7zh:-KccJkaZSH2yeilFy7zi:bcabg77) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope_output.cc#L11)):*
> ```cpp
> namespace sensors {
> 
> const int GyroscopeOutputConstants::kRollIndex;
> ```

i would switch to wx wy wz. rpy might imply something more than what you are doing. 

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccL1B1AXJ8cQmX1dNV:-KccL1B1AXJ8cQmX1dNW:bmjirn3) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/test/gyroscope_test.cc#L80)):*
> ```cpp
> // Tests Gyroscope's various accessor and streaming to-string methods.
> TEST_F(TestGyroscope, AccessorsAndToStringTest) {
>   // RigidBodyFrame<double> frame("gyroscope frame", &tree_->world(),
> ```

dead code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@siyuanfeng-tri, I addressed your comments. PTAL. Thanks!

---

Review status: 3 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccIWGAePOy3WtuB_Kg:-KccRZUg9UVwuLEY35wp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.cc#L61)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

no translational acc? 
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 29 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccImD03mgsGS93zmoz:-KccS3vgsB0GJMLnJJD2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L29)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

G's rotational velocity, Go is a point, and it is kinda weird to say it angular velocity. Also everything rigidly attached to G has the same angular velocity. 
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 31 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJ1bvaK96m3CQxvcH:-KccS4PUGjwDh1DYgey7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L31)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

G
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJJi9jgsOGZw-EheA:-KccSGfEt-CxC3o6h4tX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope.h#L79)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, should we switch to drake_copyable? there is a macro for something like this. 
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope_output.cc, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccJkaZSH2yeilFy7zh:-KccSUq1ygblvBSRZQlS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/gyroscope_output.cc#L11)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i would switch to wx wy wz. rpy might imply something more than what you are doing. 
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccL1B1AXJ8cQmX1dNV:-KccSa0Hr5Fb_P98MC5P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cbad34992a45a58f37a5c41495465e5dcf9f9e73/drake/systems/sensors/test/gyroscope_test.cc#L80)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

dead code?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KccSb0A7z8lkeYnL1NW:b6giek)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks liang!

---

Reviewed 4 of 5 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccXOgSoFlpa6J1mtE_:-KccXOgTZOV4Ej4U1rdF:b-n1mgdl) ([raw file](https://github.com/robotlocomotion/drake/blob/17888741c7a195e8c5646ee687f5d421adf29c22/drake/systems/sensors/gyroscope.cc#L66)):*
> ```cpp
>   // Computes ω_G, the gyroscope's rotational velocity in the gyroscope's
>   // frame.
>   const auto w_VGo_G = R_GW * w_WGo_W;
> ```

you probably mean w_WGo_G (angular vel of G measrued in W expressed in G), but i'd just change it to w_WG_G

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KccXpRIRO0BURMi3uLk:b-7xgx90)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks @siyuanfeng-tri.

+@soonho-tri and +@jwnimmer-tri for platform review please. Thanks!

---

Review status: 6 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccXOgSoFlpa6J1mtE_:-KccZjoCkwSuMz42glUB:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/17888741c7a195e8c5646ee687f5d421adf29c22/drake/systems/sensors/gyroscope.cc#L66)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

you probably mean w_WGo_G (angular vel of G measrued in W expressed in G), but i'd just change it to w_WG_G
</blockquote></details>

Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KccZes0g6pa7W7I5cUo:b-7fttk5)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
check-point.

---

Reviewed 3 of 8 files at r2, 3 of 5 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/sensors/BUILD, line 104 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccbDYGtF9hIc5ZKr5z:-KccbDYHZJckTPa31Apm:b-yi37r8) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/BUILD#L104)):*
> ```Python
>     ],
> )
> ```

nitpick: let's zap a line here.

---

*[drake/systems/sensors/gyroscope.cc, line 3 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcci6MfqmoYqKV_oPno:-Kcci6MfqmoYqKV_oPnp:b-gpqvg2) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L3)):*
> ```cpp
> #include "drake/systems/sensors/gyroscope.h"
> 
> #include "drake/math/quaternion.h"
> ```

I guess this header is not being used in the file?

---

*[drake/systems/sensors/gyroscope.cc, line 34 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccehfpLczyuW1kFJEx:-KccehfpLczyuW1kFJEy:bei45rw) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L34)):*
> ```cpp
> 
> void Gyroscope::DoCalcOutput(const systems::Context<double>& context,
>                                  systems::SystemOutput<double>* output) const {
> ```

Bad indentation; please consider clang-format.

---

*[drake/systems/sensors/gyroscope.cc, line 69 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchWCiSPMGeOy5JUxp:-KcchWCiSPMGeOy5JUxq:bwh3lho) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L69)):*
> ```cpp
> 
>   // Saves the angular velocity values into the output port.
>   BasicVector<double>* output_vector =
> ```

nitpick: ` BasicVector<double>* const` compiles as well.

---

*[drake/systems/sensors/gyroscope.h, line 33 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccfy7X2PcEbxIObELp:-Kccfy7X2PcEbxIObELq:b-97x6au) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L33)):*
> ```objc
> ///  - `R_GW` be the rotation matrix from `W` to `G`.
> ///  - `ω_G_G` be `G`'s rotational velocity expressed in `G`. This is the output
> ///     output of the gyroscope.
> ```

Duplicated `output` in the comment.

---

*[drake/systems/sensors/gyroscope.h, line 46 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccevpMhmOpF8ACLbEF:-KccevpNJSA93hHWcuX1:bi0o01t) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L46)):*
> ```objc
> /// accessor:
> ///
> ///  - get_input_port(): Contains the state (i.e., postion and velocity)
> ```

Is postion a typo of position?

---

*[drake/systems/sensors/gyroscope.h, line 110 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcceVqym0s-CF_MBjTE:-KcceVqym0s-CF_MBjTF:bei45rw) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L110)):*
> ```objc
>   /// this output vector.
>   std::unique_ptr<BasicVector<double>> AllocateOutputVector(
>       const OutputPortDescriptor<double>& descriptor) const override;
> ```

Bad indentation; please consider clang-format.

---

*[drake/systems/sensors/gyroscope_output.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcciaWYZJ4TVm8uO0S-:-KcciaWYZJ4TVm8uO0S0:b-bagoh1) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L22)):*
> ```objc
>   /// Default constructor.  Sets all rows to zero.
>   GyroscopeOutput();
> ```

BTW, please consider using `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN` here.

@jwnimmer-tri , what's our policy regarding this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KcckFEYJ2JByU9O3T4_:b5ba6iw)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/sensors/gyroscope.h, line 110 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcceVqym0s-CF_MBjTE:-Kccmj1dy9meAESQtGny:byxa3q6) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L110)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Bad indentation; please consider clang-format.
</blockquote></details>

This instance I think is allowed by cppguide in this shape?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/sensors/gyroscope_output.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcciaWYZJ4TVm8uO0S-:-Kccmq8yuRLs_SaRDJ2K:b-dnufzx) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L22)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, please consider using `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN` here.

@jwnimmer-tri , what's our policy regarding this?
</blockquote></details>

The defect is "Missing disable (`= delete`) for copy / move / assign."  Whether the author writes it out longhand or uses the macro is currently discretionary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.h, line 80 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccoj3MiWjIWocoyktl:-Kccoj3MiWjIWocoyktm:bei45rw) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L80)):*
> ```objc
>   ///
>   Gyroscope(const std::string& name, const RigidBodyFrame<double>& frame,
>       const RigidBodyTree<double>& tree);
> ```

Bad indentation; please consider clang-format.

---

*[drake/systems/sensors/gyroscope.h, line 110 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcceVqym0s-CF_MBjTE:-KccokXjEP0URk15OKy0:bc5egtq) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This instance I think is allowed by cppguide in this shape?
</blockquote></details>

@jwnimmer-tri , you're right. I meant the one above. My bad.

---

*[drake/systems/sensors/gyroscope_output.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcciaWYZJ4TVm8uO0S-:-KccopqcJwg4NMMeagyA:b-1xw4kg) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The defect is "Missing disable (`= delete`) for copy / move / assign."  Whether the author writes it out longhand or uses the macro is currently discretionary.
</blockquote></details>

@jwnimmer-tri , I see. Thank you!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/sensors/BUILD, line 104 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccbDYGtF9hIc5ZKr5z:-Kccn0zjKwnDdPWqgJy-:b-k0y4ji) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/BUILD#L104)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

nitpick: let's zap a line here.
</blockquote></details>

Agreed.  FYI, we can run `tools/buildifier.sh drake/systems/sensors/BUILD` and it will fix this double-blank line (and any other formatting errors).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccXOgSoFlpa6J1mtE_:-KccpfGz1MeFKKCswSkg:b-kznlx) ([raw file](https://github.com/robotlocomotion/drake/blob/17888741c7a195e8c5646ee687f5d421adf29c22/drake/systems/sensors/gyroscope.cc#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done. Thanks.
</blockquote></details>

BTW, just a reminder that angular velocity is not associated with a point; it is a property of a frame. So it isn't correct to write `w_WGo_W` (where Go is G's origin). So as @siyuanfeng-tri said, that should only be `w_WG_W`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@soonho-tri, thanks for the partial review. I've addressed your comments.

---

Review status: 4 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/sensors/BUILD, line 104 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccbDYGtF9hIc5ZKr5z:-Kccn_lVB7z7kkvDF1JK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/BUILD#L104)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

nitpick: let's zap a line here.
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccXOgSoFlpa6J1mtE_:-KccqJqPzQoUIs8jdKj1:bwymlx3) ([raw file](https://github.com/robotlocomotion/drake/blob/17888741c7a195e8c5646ee687f5d421adf29c22/drake/systems/sensors/gyroscope.cc#L66)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, just a reminder that angular velocity is not associated with a point; it is a property of a frame. So it isn't correct to write `w_WGo_W` (where Go is G's origin). So as @siyuanfeng-tri said, that should only be `w_WG_W`.
</blockquote></details>

Thanks for the reminder!

---

*[drake/systems/sensors/gyroscope.cc, line 3 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcci6MfqmoYqKV_oPno:-KccnfBa_wwH2VQutkbV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L3)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I guess this header is not being used in the file?
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.cc, line 34 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccehfpLczyuW1kFJEx:-KccntwjG5Y9F7Dn4Azq:b-vlx86g) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L34)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Bad indentation; please consider clang-format.
</blockquote></details>

Done. If desired, I'll submit a follow-up PR that contains only clang-format changes.

---

*[drake/systems/sensors/gyroscope.cc, line 69 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchWCiSPMGeOy5JUxp:-KccoOTHLFOr_DKzPc_V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.cc#L69)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

nitpick: ` BasicVector<double>* const` compiles as well.
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 33 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccfy7X2PcEbxIObELp:-Kccp_FvsaB46EwSpve2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L33)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Duplicated `output` in the comment.
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 46 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccevpMhmOpF8ACLbEF:-Kccpd_DkOAAg81TPh12:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L46)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Is postion a typo of position?
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 80 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccoj3MiWjIWocoyktl:-KccpiZsgkFZd30zK8dY:bodkhna) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L80)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Bad indentation; please consider clang-format.
</blockquote></details>

Done. I'll submit a followup PR with just clang-format changes.

---

*[drake/systems/sensors/gyroscope_output.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcciaWYZJ4TVm8uO0S-:-KccqE86RAA9Vtr1c4ND:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L22)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@jwnimmer-tri , I see. Thank you!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KccqSjpN2iRZp_6dquw:b-4lxe16)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcd0TXHpv5vhZDcD4Rd:-Kcd0TXHpv5vhZDcD4Re:b7gkxsc) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/test/gyroscope_test.cc#L37)):*
> ```cpp
>   TestGyroscope() : tree_(make_unique<RigidBodyTree<double>>()) {}
> 
>   void SetUp() {
> ```

Missing `override`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-Kcd0Vz30k3y65bQypr7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 37 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcd0TXHpv5vhZDcD4Rd:-Kcd3ErL76hPXA_1XlgP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/test/gyroscope_test.cc#L37)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Missing `override`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 8 files at r2, 3 of 5 files at r3, 1 of 1 files at r4, 4 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.h, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccinsY2j430MdSHVDI:-KccinsZyMw8XcsR5lpV:b-ssj0tq) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L6)):*
> ```objc
> #include <memory>
> #include <string>
> #include <vector>
> ```

I think perhaps many of the above are unused?

---

*[drake/systems/sensors/gyroscope.h, line 112 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchbIyv7W7dMEMbdY4:-KcchbIyv7W7dMEMbdY5:blbc3de) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L112)):*
> ```objc
>       const OutputPortDescriptor<double>& descriptor) const override;
> 
>   friend std::ostream& operator<<(std::ostream& out,
> ```

_"Overload << as a streaming operator for your type only if your type represents a value, and << writes out a human-readable string representation of that value. Avoid exposing implementation details in the output of <<; if you need to print object internals for debugging, use named functions instead (a method named DebugString() is the most common convention)."_

This use is borderline, but I think likely does not meet the standards of cppguide.

---

*[drake/systems/sensors/gyroscope.h, line 113 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchEAPFYoPRFw0sYig:-KcchEAPFYoPRFw0sYih:b2tndzj) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L113)):*
> ```objc
> 
>   friend std::ostream& operator<<(std::ostream& out,
>                                   const Gyroscope& depth_sensor);
> ```

Typo s/depth_sensor/sensor/

---

*[drake/systems/sensors/gyroscope_output.cc, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccgdQoKckxQW1HXi0B:-KccgdQoKckxQW1HXi0C:b-6fgs8i) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.cc#L24)):*
> ```cpp
>   // TODO(liang.fok) Optimize the following code by re-interpreting the three
>   // rotational velocity values, which are already sitting sequentially in
>   // BasicVector's memory, as a Vector3<T> with no copying.
> ```

BTW I'm not sure this comment adds any value?  Any code might end up being a hot spot and need to be optimized.  Fixing this TODO would require changing the method signature, and I am also skeptical that the performance impact is meaningful.

---

*[drake/systems/sensors/gyroscope_output.h, line 9 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccdu-1a-f8eq7zUheD:-Kccdu-2iyCshmY5_0u3:b-iyo1cb) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L9)):*
> ```objc
> namespace sensors {
> 
> struct GyroscopeOutputConstants {
> ```

Unclear why this is public?  Only the `cc` file seems to use it.

If it remains public, then missing API documentation.

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 161 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcd9Zq_0gkhKkVmFg1o:-Kcd9Zq_0gkhKkVmFg1p:b-xj92me) ([raw file](https://github.com/robotlocomotion/drake/blob/6046509df6a59de4426628402593c551debbe537/drake/systems/sensors/test/gyroscope_test.cc#L161)):*
> ```cpp
> }  // namespace sensors
> }  // namespace systems
> }  // namespace drake
> ```

I think `get_rotational_velocities` should be (lightly) tested, and I don't see it anywhere here though?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/gyroscope.h, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccinsY2j430MdSHVDI:-KcdGfzQbaGPQU_dCT2-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think perhaps many of the above are unused?
</blockquote></details>

Done.

---

*[drake/systems/sensors/gyroscope.h, line 112 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchbIyv7W7dMEMbdY4:-KcdGxkvO1WoirhQptCX:b-857eya) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_"Overload << as a streaming operator for your type only if your type represents a value, and << writes out a human-readable string representation of that value. Avoid exposing implementation details in the output of <<; if you need to print object internals for debugging, use named functions instead (a method named DebugString() is the most common convention)."_

This use is borderline, but I think likely does not meet the standards of cppguide.
</blockquote></details>

Alright, I will remove it. I wanted to include the current measurement in the string but that's not easy to do in the System Framework.

---

*[drake/systems/sensors/gyroscope.h, line 113 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchEAPFYoPRFw0sYig:-KcdHm5O_Lv9emUBqr8X:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Typo s/depth_sensor/sensor/
</blockquote></details>

Removed.

---

*[drake/systems/sensors/gyroscope_output.cc, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KccgdQoKckxQW1HXi0B:-KcdHtTvjjuSEJcyMqJa:bxdc7fw) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.cc#L24)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I'm not sure this comment adds any value?  Any code might end up being a hot spot and need to be optimized.  Fixing this TODO would require changing the method signature, and I am also skeptical that the performance impact is meaningful.
</blockquote></details>

Done. (Removed.)

---

*[drake/systems/sensors/gyroscope_output.h, line 9 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kccdu-1a-f8eq7zUheD:-KcdIMcyY7A2PqpakdA6:b-fbpky) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope_output.h#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why this is public?  Only the `cc` file seems to use it.

If it remains public, then missing API documentation.
</blockquote></details>

I envision it will be used by downstream systems and users to understands the semantics of each index within `GyroscopeOutput`. I've added documentation.

---

*[drake/systems/sensors/test/gyroscope_test.cc, line 161 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-Kcd9Zq_0gkhKkVmFg1o:-KcdMGrWclEixoyhyMKi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/6046509df6a59de4426628402593c551debbe537/drake/systems/sensors/test/gyroscope_test.cc#L161)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think `get_rotational_velocities` should be (lightly) tested, and I don't see it anywhere here though?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/gyroscope.h, line 112 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-KcchbIyv7W7dMEMbdY4:-KcdXJx-b5x5LewP7ydO:b-hozlk3) ([raw file](https://github.com/robotlocomotion/drake/blob/5be9a51826379dab10bbf2a2e9218c102b6df233/drake/systems/sensors/gyroscope.h#L112)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alright, I will remove it. I wanted to include the current measurement in the string but that's not easy to do in the System Framework.
</blockquote></details>

FYI Renaming to `DebugString` would also have been fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4795#-:-KcdXe44uGXxfWGtAPCU:bnfp4nl)*
<!-- Sent from Reviewable.io -->
